File: PythonAlgorithms/MergeQuickSorts.py
import random
import cProfile
import pstats
import io
import sys
from memory_profiler import profile


def mergeSortStart(lst):
    return mergeSort(lst, 0, len(lst) - 1)

@profile
def mergeSort(lst, first, last):
    middle = 0
    count = 0
    if first < last:
        middle = (first + last)//2
        count += mergeSort(lst, first, middle)
        count += mergeSort(lst, middle+1, last)
        count += mergeLists(lst, first, middle, middle+1, last)
    return count


def mergeLists(lst, start1, end1, start2, end2):
    count = 0
    result = [None] * len(lst)
    finalStart = start1
    finalEnd = end2
    indexC = 0
    while start1 <= end1 and start2 <= end2:
        count += 1
        if lst[start1] < lst[start2]:
            result[indexC] = lst[start1]
            start1 += 1
        else:
            result[indexC] = lst[start2]
            start2 += 1
        
        indexC += 1
    
    if start1 <= end1:
        for i in range(start1, end1+1):
            result[indexC] = lst[i]
            indexC += 1
    else:
       for i in range(start2, end2+1):
            result[indexC] = lst[i]
            indexC += 1 
    
    indexC = 0
    for i in range(finalStart, finalEnd+1):
        lst[i] = result[indexC]
        indexC += 1
    
    return count

def pivotArr(pivot, arr, left, right, comparisons):
    temp = arr[right]
    arr[right]= arr[pivot]
    arr[pivot] = temp
    i = left
    for j in range(i, right):
            comparisons[0] += 1
            if arr[j] < arr[right]:
                temp = arr[i]
                arr[i] = arr[j]
                arr[j] = temp
                i += 1
    
    temp = arr[i]
    arr[i] = arr[right]
    arr[right] = temp
    return i

@profile
def quickSortPivotFirst(lst, first, last, comparisons):
    count = 0
    pivot = 0
    if first < last:
        pivot = pivotArr(first, lst, first, last, comparisons)
        quickSortPivotFirst(lst, first, pivot - 1, comparisons)
        quickSortPivotFirst(lst, pivot + 1, last, comparisons)
    return count

def quickSortPivotLast(lst, first, last, comparisons):
    count = 0
    pivot = 0
    if first < last:
        pivot = pivotArr(last, lst, first, last, comparisons)
        quickSortPivotLast(lst, first, pivot - 1, comparisons)
        quickSortPivotLast(lst, pivot + 1, last, comparisons)
    return count

def quickSortPivotRandom(lst, first, last, comparisons):
    count = 0
    pivot = 0
    if first < last:
        pivot = pivotArr(random.randint(first, last), lst, first, last, comparisons)
        quickSortPivotRandom(lst, first, pivot - 1, comparisons)
        quickSortPivotRandom(lst, pivot + 1, last, comparisons)
    return count


# Tests:


def test_quickSort(count = 0):
    toSort = random.sample(range(0, 100000), 10000)
    comparisons = [0]
    
    profiler = cProfile.Profile()
    profiler.enable()

    quickSortPivotFirst(toSort, 0, len(toSort) - 1, comparisons)

    profiler.disable()
    profiler.dump_stats("./PythonAlgorithms/outputs/quickSort/quickSort_py"+str(count)+".stats")

    # Copy to CSV
    result = io.StringIO()
    pstats.Stats(profiler,stream=result).print_stats()
    result=result.getvalue()
    # chop the string into a csv-like buffer
    result='ncalls'+result.split('ncalls')[-1]
    result='\n'.join([','.join(line.rstrip().split(None,5)) for line in result.split('\n')])

    # save it to disk
 
    with open("./PythonAlgorithms/outputs/quickSort/quickSort_py"+str(count)+".csv", 'w+') as f:
        #f=open(result.rsplit('.')[0]+'.csv','w')
        f.write(result)
        f.close()

def test_mergeSort(count = 0):
    toSort = random.sample(range(0, 100000), 10000)

    profiler = cProfile.Profile()
    profiler.enable()

    mergeSort(toSort, first = 0, last = len(toSort) - 1)

    profiler.disable()
    profiler.dump_stats("./PythonAlgorithms/outputs/mergeSort/mergeSort_py"+str(count)+".stats")

    # Copy to CSV
    result = io.StringIO()
    pstats.Stats(profiler,stream=result).print_stats()
    result=result.getvalue()
    # chop the string into a csv-like buffer
    result='ncalls'+result.split('ncalls')[-1]
    result='\n'.join([','.join(line.rstrip().split(None,5)) for line in result.split('\n')])

    # save it to disk

    with open('./PythonAlgorithms/outputs/mergeSort/mergeSort_py'+str(count)+'.csv', 'w+') as f:
        #f=open(result.rsplit('.')[0]+'.csv','w')
        f.write(result)
        f.close()

def main(count = 0):
    test_mergeSort(count)
    test_quickSort(count)

if __name__ == "__main__":
    main(int(sys.argv[1]))

File: JavaAlgorithms/MergeQuickSorts.java
package JavaAlgorithms;

/** ／(^ㅅ^)＼
 * @file JavaAlgorithms.Matrix.java
 * @author Kali Hale
 * @version 1.0
 *
 * Code from CS 310 taught by Dr. Craig Reinhart at California Lutheran University,
 * used with permission.
 */

import java.io.FileWriter;
import java.io.IOException;
import java.util.Arrays;
import java.util.Random;

public class MergeQuickSorts
{
    public static void main(String[] args)
    {
        try
        {
            FileWriter csvwriter = new FileWriter("quicksortPivot1RandNums.csv");
            Random random = new Random(37);

            int[] list;
            int[] comparisons = new int[1];

            for (int i = 10; i <= 1000000; i *= 10)
            {
                list = new int[i];

                for (int j = 0; j < i; j++)
                {
                    list[j] = random.nextInt(100000);
                }
                System.out.println(Arrays.toString(list));
                quickSortPivotFirst(list, 0, list.length - 1, comparisons);
                csvwriter.write(i + "," + comparisons[0] + "\n");
                System.out.println(Arrays.toString(list));
            }
            csvwriter.close();
        }catch(IOException e)
        {
            System.out.println(e);
        }
    }
    public static int mergeSort(int[] list, int first, int last)
    {
        int middle;
        int count = 0;
        if(first < last)
        {
            middle = (first + last)/2;
            count += mergeSort(list, first, middle);
            count += mergeSort(list, middle+1, last);
            count += mergeLists(list, first, middle, middle+1, last);
        }
        return count;
    }
    public static int mergeLists(int[] list, int start1, int end1, int start2, int end2)
    {
        int count = 0;
        int[] result = new int[list.length];
        int finalStart = start1;
        int finalEnd = end2;
        int indexC = 0;
        while(start1 <= end1 && start2 <= end2)
        {
            count++;
            if(list[start1] < list[start2])
            {
                result[indexC] = list[start1];
                start1++;
            }
            else
            {
                result[indexC] = list[start2];
                start2++;
            }
            indexC++;
        }
        if(start1 <= end1)
        {
            for(int i = start1; i <= end1; i++)
            {
                result[indexC] = list[i];
                indexC++;
            }
        }
        else
        {
            for(int i = start2; i <= end2; i++)
            {
                result[indexC] = list[i];
                indexC++;
            }
        }
        indexC = 0;
        for(int i = finalStart; i <= finalEnd; i++)
        {
            list[i] = result[indexC];
            indexC++;
        }
        return count;
    }
    public static int quickSortPivotFirst(int[] list, int first, int last, int[] comparisons)
    {
        int count = 0;
        int pivot;
        if(first < last)
        {
            pivot = pivot(first, list, first, last, comparisons);
            quickSortPivotFirst(list, first, pivot - 1, comparisons);
            quickSortPivotFirst(list, pivot + 1, last, comparisons);
        }
        return count;
    }
    public static int quickSortPivotLast(int[] list, int first, int last, int[] comparisons)
    {
        int count = 0;
        int pivot;
        if(first < last)
        {
            pivot = pivot(last, list, first, last, comparisons);
            quickSortPivotLast(list, first, pivot - 1, comparisons);
            quickSortPivotLast(list, pivot + 1, last, comparisons);
        }
        return count;
    }
    public static int quickSortPivotRandom(int[] list, int first, int last, int[] comparisons)
    {
        Random random = new Random(13);
        int count = 0;
        int pivot;
        if(first < last)
        {
            pivot = pivot(random.nextInt(last - first) + first, list, first, last, comparisons);
            quickSortPivotRandom(list, first, pivot - 1, comparisons);
            quickSortPivotRandom(list, pivot + 1, last, comparisons);
        }
        return count;
    }
    public static int pivot(int pivot, int[] arr, int left, int right, int[] comparisons)
    {
        int temp = arr[right];
        arr[right] = arr[pivot];
        arr[pivot] = temp;
        int i = left;

        for(int j = i; j < right; j++)
        {
            comparisons[0]++;
            if(arr[j] < arr[right])
            {
                temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
                i++;
            }

        }
        temp = arr[i];
        arr[i] = arr[right];
        arr[right] = temp;
        return i;
    }
}
